src/imex/transfer.rs

```diff
@@ -0,0 +1,627 @@
+            .get_blobdir()
+            .parent()
+            .ok_or(anyhow!("Context dir not found"))?;
+        let dbfile = context_dir.join(DBFILE_BACKUP_NAME);
+        if fs::metadata(&dbfile).await.is_ok() {
+            fs::remove_file(&dbfile).await?;
+            warn!(context, "Previous database export deleted");
+        }
+        dbfile
+    } else {
+        ensure!(name.starts_with("blob/"), "malformatted blob name");
+        let blobname = name.rsplit('/').next().context("malformatted blob name")?;
+        context.get_blobdir().join(blobname)
+    };
+    let _guard = if name.starts_with("db/") {
+        Some(DeleteOnDrop(path.clone()))
+    } else {
+        None
+    };
+
+    let mut wrapped_reader = progress.wrap_async_read(&mut reader);
+    let file = File::create(&path).await?;
+    let mut file = BufWriter::with_capacity(128 * 1024, file);
+    io::copy(&mut wrapped_reader, &mut file).await?;
+    file.flush().await?;
+
+    if name.starts_with("db/") {
+        context
+            .sql
+            .import(&path, ticket.token.to_string())
+            .await
+            .context("cannot import database")?;
+        fs::remove_file(&path)
+            .await
+            .with_context(|| format!("database import file: {}", path.display()))?;
+    }
+    Ok(reader)
+}
+
+/// Spawns a task proxying progress events.
+///
+/// This spawns a tokio task which receives events from the [`ProgressEmitter`] and sends
+/// them to the context.  The task finishes when the emitter is dropped.
+///
+/// This could be done directly in the emitter by making it less generic.
+fn spawn_progress_proxy(context: Context, mut rx: broadcast::Receiver<u16>) {
+    tokio::spawn(async move {
+        loop {
+            match rx.recv().await {
+                Ok(step) => context.emit_event(ReceiveProgress::BlobProgress(step).into()),
+                Err(RecvError::Closed) => break,
+                Err(RecvError::Lagged(_)) => continue,
+            }
+        }
+    });
+}
+
+/// Create [`EventType::ImexProgress`] events using readable names.
+///
+/// Plus you get warnings if you don't use all variants.
+#[derive(Debug)]
+enum ReceiveProgress {
+    Connected,
+    CollectionRecieved,
+    /// A value between 0 and 85 interpreted as a percentage.
+    ///
+    /// Other values are already used by the other variants of this enum.
+    BlobProgress(u16),
+    Completed,
+    Failed,
+}
+
+impl ReceiveProgress {
+    /// The maximum value for [`ReceiveProgress::BlobProgress`].
+    ///
+    /// This only exists to keep this magic value local in this type.
+    fn max_blob_progress() -> u16 {
+        85
+    }
+}
+
+impl From<ReceiveProgress> for EventType {
+    fn from(source: ReceiveProgress) -> Self {
+        let val = match source {
+            ReceiveProgress::Connected => 50,
+            ReceiveProgress::CollectionRecieved => 100,
+            ReceiveProgress::BlobProgress(val) => 100 + 10 * val,
```

Comment:
I am confused

```rust
enum SendProgress {
    Failed = 0,
    Started = 100,
    DatabaseExported = 300,
    CollectionCreated = 400,
    ProviderListening = 500,
    ClientConnected = 600,
    TransferStarted = 650,
    TransferFinished = 950,
    Completed = 1000,
}
```
has already quite a lot of details, what exactly are you missing?

Reply:
sorry for being unclear. i meant additional events between `SendProgress::TransferStarted`  and `SendProgress::TransferFinished`. and between `SendProgress::Started` and `SendProgress::DatabaseExported`.  that would allow showing a "real" progress bar also on the sending device. 

`ReceiveProgress::BlobProgress(val)` we have 100 values which allows showing a smart progress. the sender is stuck at 650 during all the time.

Reply:
The database export does not give us any information, so there is really nothing we can print between `Started` and `DatabaseExported`.

Reply:
We currently do not have any more detailed events inside of `iroh` on the sender side for the actual transfer progress, but I will look into what it will take adding those.

Reply:
work is done here: https://github.com/n0-computer/iroh/pull/860

Reply:
Updated, you should have now a more detailed progress

Reply:
cool, thanks a lot!